add in kernel ksmbd

neheb opened the pull request:
Currently there's a conflict between the two packages. I have an idea that I don't like on how to fix it, but I'll set to draft for now.

neheb committed:
linux/modules: add ksmbd support for 5.15

5.15 is the first version to come with ksmbd. Use it instead of the
module as the linux one is better maintained.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

Ansuel commented:
@neheb an overkill alternative would be backport ksmbd to 5.10

neheb commented:
I think ksmbd oot should be kept for now as it supports CIFS 1.0. The kernel one does not.

hauke commented:
Do people still want to use  CIFS 1.0? Windows 10 deactivated it by default.

Ansuel commented:
Some old TV / system still require 1.0 IMHO it's still used a lot.

Il Dom 2 Ott 2022, 18:35 Hauke Mehrtens ***@***.***> ha
scritto:

> Do people still want to use CIFS 1.0? Windows 10 deactivated it by default.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10296#issuecomment-1264683349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQVBTHZGX3ZXWTC2HCTWBG2WDANCNFSM532U72ZQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

john-tho committed:
kernel: modules: split smbfs-common from cifs

This kernel config symbol is a shared library used for both CIFS and the
in-tree SMB_SERVER.
Split it out into kmod-fs-smbfs-common from kmod-fs-cifs, and depend on
it, so that later in-tree ksmbd may also depend on it.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

john-tho reviewed (commented):
I ran into compile errors with the external ksmbd for kernel 6.1. They were fixed by a bump + cherry pick, but I was intrigued by the in-tree option. No rush on this.
I opened a PR against my own tree (to avoid polluting OpenWrt PRs) with what I ended up with. Only compile tested (with WIP 6.1 kernel) though. https://github.com/john-tho/openwrt/pull/4
I do not use SMB at home, so cannot test with any understanding, and do not want to take over the PR. I saw it was stale, but thought it worth pursuing. Feel free to take any or none of my changes.

My eventual process was:
- split the `SMBFS_COMMON` module from cifs package, and add as a depend to `ksmbd` https://github.com/john-tho/openwrt/pull/4/commits/7af47ff9a4f79e8dc23adb308c4cef7092a865d3
- cherry pick this _in kernel ksmbd_ commit
- rename to `kmod-fs-ksmbd-external`, and `kmod-fs-ksmbd-intree`, both provide & conflict on `kmod-fs-ksmbd`, and `external` conflicts on `intree`. Not that sure about the conflicts here though. https://github.com/john-tho/openwrt/pull/4/commits/3c6d3a0e8c634a701ba33e27e41619266aee7846
- allow `SMB_SERVER` to be compiled as module & add all kmods depends (`external` deps, except `md4`) https://github.com/john-tho/openwrt/pull/4/commits/4eef7cc57aed4c0531cf632272ebeeb05f09452f

Cheers

Neustradamus commented:
To follow

neheb commented on package/kernel/ksmbd/Makefile:
```diff
@@ -23,6 +23,7 @@ define KernelPackage/fs-ksmbd
 	URL:=https://github.com/cifsd-team/cifsd
 	FILES:=$(PKG_BUILD_DIR)/ksmbd.ko
 	DEPENDS:= \
+		@LINUX_5_10 \
```
I disagree that external one is desired. In all other examples (wireguard and exfat come to mind), external module was removed and in kernel one is used.

neheb commented:
@john-tho applied some of your commits

hauke commented on package/kernel/linux/modules/fs.mk:
```diff
@@ -332,6 +348,45 @@ endef
+  CONFLICTS:=kmod-fs-ksmbd
+  DEPENDS:= \
+	    @!LINUX_5_10 \
+	    +LINUX_5_15:kmod-fs-smbfs-common \
```
Please make this dependency unconditionally, we only support kernel 5.15.
```suggestion
	    +kmod-fs-smbfs-common \
```

neheb replied:
Done. I think this is what you wanted.

john-tho replied:
I read it as: because we depend on `@!LINUX_5_10` beforehand, no need for the `LINUX_5_15` next line, but what you have done _might_ work as is, because `kmod-fs-smbfs-common` has a !5.10 depend.
Also, if we are switching external for in-tree, we should add the 5.10 dependency on external before we add the in-tree package, and you can cut the PROVIDES and CONFLICTS on the in-tree.

neheb replied:
Yeah. The @!LINUX_5_10 is an implicit dependency.

neheb replied:
Still getting these errors:

```
tmp/.config-package.in:10922:error: recursive dependency detected!
tmp/.config-package.in:10922:	symbol PACKAGE_kmod-fs-ksmbd is selected by PACKAGE_ksmbd-server
tmp/.config-package.in:90226:	symbol PACKAGE_ksmbd-server depends on PACKAGE_kmod-fs-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:56063:error: recursive dependency detected!
tmp/.config-package.in:56063:	symbol PACKAGE_luci-app-ksmbd depends on PACKAGE_luci-app-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

```

neheb commented:
```
tmp/.config-package.in:10922:error: recursive dependency detected!
tmp/.config-package.in:10922:	symbol PACKAGE_kmod-fs-ksmbd depends on PACKAGE_kmod-fs-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:90227:error: recursive dependency detected!
tmp/.config-package.in:90227:	symbol PACKAGE_ksmbd-server depends on PACKAGE_ksmbd-server
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:56064:error: recursive dependency detected!
tmp/.config-package.in:56064:	symbol PACKAGE_luci-app-ksmbd depends on PACKAGE_luci-app-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations
```

No idea how to fix. getting rid of CONFLICTS only gets me part of the way there.

neheb commented:
The issue becomes that under 5.15, both external module and internal get selected.

john-tho commented:
Ah, I guess if the packages have the same name, cannot tell them apart properly?
Sorry. I did not try that, when straight to rename both, with `PROVIDES` & `CONFLICTS`

neheb commented:
I tried the following:

```
--- a/package/kernel/ksmbd/Makefile
+++ b/package/kernel/ksmbd/Makefile
@@ -17,11 +17,13 @@ include $(INCLUDE_DIR)/package.mk
 TAR_OPTIONS+= --strip-components 1
 TAR_CMD=$(HOST_TAR) -C $(1) $(TAR_OPTIONS)
 
-define KernelPackage/fs-ksmbd
+define KernelPackage/fs-ksmbd-oot
        SUBMENU:=Filesystems
        TITLE:=SMB kernel server support
        URL:=https://github.com/cifsd-team/cifsd
        FILES:=$(PKG_BUILD_DIR)/ksmbd.ko
+       PROVIDES:=kmod-fs-ksmbd
+       CONFLICTS:=kmod-fs-ksmbd
        DEPENDS:= \
                @LINUX_5_10 \
                +kmod-nls-base \
@@ -44,15 +46,15 @@ endef
 # The last two DEPENDS are hacks in order to get CONFIG_ASN1 and CONFIG_OID_REGISTRY
 # which it seems can't be selected independently. Some bug in either base or upstream.
 
-define KernelPackage/fs-ksmbd/description
+define KernelPackage/fs-ksmbd-oot/description
   Ksmbd is an In-kernel SMBv(1)2/3 fileserver.
   It's an implementation of the SMB protocol in kernel space for sharing files and IPC services over network.
 endef
 
-define KernelPackage/fs-ksmbd/config
+define KernelPackage/fs-ksmbd-oot/config
 config KSMBD_SMB_INSECURE_SERVER
        bool "Support for insecure SMB1/CIFS and SMB2.0 protocols"
-       depends on PACKAGE_kmod-fs-ksmbd
+       depends on PACKAGE_kmod-fs-ksmbd-oot
        help
                This enables deprecated insecure protocols dialects: SMB1/CIFS and SMB2.0.
        default y
@@ -71,4 +73,4 @@ define Build/Compile
        modules
 endef
 
-$(eval $(call KernelPackage,fs-ksmbd))
+$(eval $(call KernelPackage,fs-ksmbd-oot))
diff --git a/package/kernel/linux/modules/fs.mk b/package/kernel/linux/modules/fs.mk
index 2f8942e71c..cb7faff805 100644
--- a/package/kernel/linux/modules/fs.mk
+++ b/package/kernel/linux/modules/fs.mk
@@ -352,8 +352,7 @@ $(eval $(call KernelPackage,fs-jfs))
 define KernelPackage/fs-ksmbd
   SUBMENU:=$(FS_MENU)
   TITLE:=SMB kernel server support
-  PROVIDES:=kmod-fs-ksmbd
-  CONFLICTS:=kmod-fs-ksmbd
+  CONFLICTS:=kmod-fs-ksmbd-oot
   DEPENDS:= \
            +kmod-fs-smbfs-common \
            +kmod-nls-base \
```

```
tmp/.config-package.in:10922:error: recursive dependency detected!
tmp/.config-package.in:10922:	symbol PACKAGE_kmod-fs-ksmbd depends on PACKAGE_kmod-fs-ksmbd-oot
tmp/.config-package.in:10947:	symbol PACKAGE_kmod-fs-ksmbd-oot depends on PACKAGE_kmod-fs-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:90228:error: recursive dependency detected!
tmp/.config-package.in:90228:	symbol PACKAGE_ksmbd-server depends on PACKAGE_ksmbd-server
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:56065:error: recursive dependency detected!
tmp/.config-package.in:56065:	symbol PACKAGE_luci-app-ksmbd depends on PACKAGE_luci-app-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"
```

neheb commented:
```
--- a/package/kernel/ksmbd/Makefile
+++ b/package/kernel/ksmbd/Makefile
@@ -17,11 +17,12 @@ include $(INCLUDE_DIR)/package.mk
 TAR_OPTIONS+= --strip-components 1
 TAR_CMD=$(HOST_TAR) -C $(1) $(TAR_OPTIONS)
 
-define KernelPackage/fs-ksmbd
+define KernelPackage/fs-ksmbd-oot
        SUBMENU:=Filesystems
        TITLE:=SMB kernel server support
        URL:=https://github.com/cifsd-team/cifsd
        FILES:=$(PKG_BUILD_DIR)/ksmbd.ko
+       PROVIDES:=kmod-fs-ksmbd
        DEPENDS:= \
                @LINUX_5_10 \
                +kmod-nls-base \
@@ -44,15 +45,15 @@ endef
 # The last two DEPENDS are hacks in order to get CONFIG_ASN1 and CONFIG_OID_REGISTRY
 # which it seems can't be selected independently. Some bug in either base or upstream.
 
-define KernelPackage/fs-ksmbd/description
+define KernelPackage/fs-ksmbd-oot/description
   Ksmbd is an In-kernel SMBv(1)2/3 fileserver.
   It's an implementation of the SMB protocol in kernel space for sharing files and IPC services over network.
 endef
 
-define KernelPackage/fs-ksmbd/config
+define KernelPackage/fs-ksmbd-oot/config
 config KSMBD_SMB_INSECURE_SERVER
        bool "Support for insecure SMB1/CIFS and SMB2.0 protocols"
-       depends on PACKAGE_kmod-fs-ksmbd
+       depends on PACKAGE_kmod-fs-ksmbd-oot
        help
                This enables deprecated insecure protocols dialects: SMB1/CIFS and SMB2.0.
        default y
@@ -71,4 +72,4 @@ define Build/Compile
        modules
 endef
 
-$(eval $(call KernelPackage,fs-ksmbd))
+$(eval $(call KernelPackage,fs-ksmbd-oot))
diff --git a/package/kernel/linux/modules/fs.mk b/package/kernel/linux/modules/fs.mk
index 2f8942e71c..be59b185ed 100644
--- a/package/kernel/linux/modules/fs.mk
+++ b/package/kernel/linux/modules/fs.mk
@@ -352,8 +352,6 @@ $(eval $(call KernelPackage,fs-jfs))
 define KernelPackage/fs-ksmbd
   SUBMENU:=$(FS_MENU)
   TITLE:=SMB kernel server support
-  PROVIDES:=kmod-fs-ksmbd
-  CONFLICTS:=kmod-fs-ksmbd
   DEPENDS:= \
            +kmod-fs-smbfs-common \
            +kmod-nls-base \
```

gets me farther but still not good enough:

```
tmp/.config-package.in:90226:error: recursive dependency detected!
tmp/.config-package.in:90226:	symbol PACKAGE_ksmbd-server depends on PACKAGE_ksmbd-server
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:56063:error: recursive dependency detected!
tmp/.config-package.in:56063:	symbol PACKAGE_luci-app-ksmbd depends on PACKAGE_luci-app-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"
```

neheb commented:
it seems the external module must be renamed. otherwise there's a conflict.

john-tho commented:
Okay, I only saw the error once I added the external DEPEND 5.10

I looked up what luci-app-firewall does for firewall3/4 now:
Looks like I should not have conflicted on the provided package. Sorry for the runaround.
bad here:
```
PROVIDES:=kmod-fs-ksmbd
CONFLICTS:=kmod-fs-ksmbd
```
Seems okay for me if both are renamed, both use PROVIDES, only external CONFLICTS on intree, and external DEPENDS 5.10: https://github.com/john-tho/openwrt/pull/4/files
that way there is not both SELECT & DEPEND in tmp/.config-package.in, and ksmbd-server depends on (intree||external)

Then, from my minimal config, I can do the following:
```
echo CONFIG_PACKAGE_kmod-fs-ksmbd-intree=m>>.config
echo CONFIG_PACKAGE_ksmbd-server=m>>.config
echo CONFIG_PACKAGE_luci-app-ksmbd=m>>.config
make defconfig
```

neheb commented:
applied your branch. I still see

```
tmp/.config-package.in:90227:error: recursive dependency detected!
tmp/.config-package.in:90227:	symbol PACKAGE_ksmbd-server depends on PACKAGE_ksmbd-server
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:56064:error: recursive dependency detected!
tmp/.config-package.in:56064:	symbol PACKAGE_luci-app-ksmbd depends on PACKAGE_luci-app-ksmbd
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

```

john-tho commented:
Sorry again, my stale config somehow…

If I have this in either the packages feed ksmbd-tools`ksmbd-server`, or in a hidden `fs-ksmbd` package, it _might_ avoid the recursive dependency? https://github.com/john-tho/openwrt/pull/4/commits/bf07f5c1c82e64a93995fb986af85bc112c09341
```
DEPENDS:= \
+LINUX_5_10:kmod-fs-ksmbd-external \
+(LINUX_5_15||LINUX_6.1):kmod-fs-ksmbd-intree
```
```
grep -n -E -A6 'config PACKAGE.*fs-ksmbd$' tmp/.config-package.in
10930:  config PACKAGE_kmod-fs-ksmbd
10931-          tristate 
10932-          default y if DEFAULT_kmod-fs-ksmbd
10933-          select PACKAGE_kmod-fs-ksmbd-external if LINUX_5_10
10934-          depends on !((LINUX_5_15||LINUX_6.1)) || !LINUX_5_10
10935-          select PACKAGE_kmod-fs-ksmbd-intree if (LINUX_5_15||LINUX_6.1)
10936-          help
```

neheb commented:
Sounds weird to have to require that though.

PolynomialDivision commented on package/kernel/linux/modules/fs.mk:
```diff
@@ -341,6 +341,27 @@ endef
+  TITLE:=SMB kernel server support
+  DEPENDS:=+kmod-asn1-decoder +kmod-oid-registry @!LINUX_5_10
+  KCONFIG:= \
+	CONFIG_SMB_SERVER=y \
```
ping @neheb

neheb committed:
ksmbd: remove

In tree ksmbd is available with 5.15 and is better updated.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
Rebased. Let's see if CI likes it.

john-tho commented:
Missing most of the ksmbd server dependencies again.
https://elixir.bootlin.com/linux/v5.15.111/source/fs/ksmbd/server.c#L622
```
MODULE_SOFTDEP("pre: ecb");
MODULE_SOFTDEP("pre: hmac");
MODULE_SOFTDEP("pre: md4");
MODULE_SOFTDEP("pre: md5");
MODULE_SOFTDEP("pre: nls");
MODULE_SOFTDEP("pre: aes");
MODULE_SOFTDEP("pre: cmac");
MODULE_SOFTDEP("pre: sha256");
MODULE_SOFTDEP("pre: sha512");
MODULE_SOFTDEP("pre: aead2");
MODULE_SOFTDEP("pre: ccm");
MODULE_SOFTDEP("pre: gcm");
MODULE_SOFTDEP("pre: crc32");
```

neheb commented:
@john-tho fixed

hauke commented:
I would like to merge this even if we loose support for SMB1. ksmbd gets many security updates and maintaining the external kernel module is not so easy, if we use the in kernel version we get the fixes for free in the minor kernel version update. @Ansuel  what do you think?
